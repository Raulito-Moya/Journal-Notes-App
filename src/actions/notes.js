import Swal from 'sweetalert2';

import { db } from "../firebase/firebase-config";
import { fileUpload } from '../helpers/fileUpload';
import { loadNotes } from "../helpers/loadNotes";
import { types } from "../types/types";




export const startNewNote =() => {
    return async (dispatch, getState) =>{  //getState es una funcione para obtener el sstate

        const uid = getState().auth.uid //obtnego el uid
       // console.log(uid);

       const newNote = {
           title: '',
           body: '',
           date: new Date().getTime() //me va adar el momento exacto en que la persona crea esta nota
       }


    const doc = await db.collection(`${ uid }/journal/notes`).add(newNote);
  
    // console.log(doc);

     dispatch( activeNote( doc.id, newNote) )
     dispatch( addNewNote( doc.id, newNote) )

    }


};


export const activeNote = (id, note) => ({ //para marcar la nota activa
      type: types.notesActive,
      payload: {
          id,
          ...note
      }
 
});

export const addNewNote = ( id, note ) => ({ //agregar la nueva nota al sidebar
     type: types.notesAddNew,
     payload:{
      id,
      ...note
     }

})


export const startLoadingNotes = ( uid ) => {
     return async(dispatch) =>{
 
        const notes = await loadNotes(uid)  //vamos a cargar las notas  de la base de datos
        dispatch(setNotes(notes))
     }
  

}

export const setNotes = ( notes ) => ({
    type: types.notesLoad,
    payload: notes
})



export const startSaveNote = (note) => { //para guardar en la base de datos
   return async(dispatch,getState) => {
        //console.log(note);
        const {uid} = getState().auth;

       if( !note.url ){
           delete note.url; //borro la propiedad url
       }

        const noteToFirestore = {...note};
        delete noteToFirestore.id;
        await db.doc(`${uid}/journal/notes/${note.id}`).update(noteToFirestore) //se guarda la nota
     
   
        dispatch(refreshNote( note.id, noteToFirestore ));
         Swal.fire('Saved', note.title, 'success');
       }

}

export const refreshNote = ( id, note ) => ({  //para cargar la accion especifica en el menu ya que si se hace con startLoadingNotes se estarian cargando todas
  type: types.notesUpdated,
  payload:{
     id, 
     note: {
         id,
         ...note
     }
   }
})



export const startUpLoading = ( file ) => {
  return async( dispatch, getState ) =>{
     
     const {active:activeNote} = getState().notes;
 
    // console.log(file);
   //  console.log(activeNote);

    Swal.fire({  //va a disparar el cartel de cargando
        title:'Uploading...',
        text:'Please wait...',
        allowOutsideClick: false,
        onBeforeOpen: ()=> {
            Swal.showLoading();
        }
    });

     const fileUrl = await fileUpload(file)
     activeNote.url = fileUrl //le actualizo la url

     console.log(fileUrl);

     dispatch(startSaveNote(activeNote))

     Swal.close();
  }

}


export const startDeleting = (id) => {
     return async(dispatch, getState) =>{

        const uid = getState().auth.uid
        await db.doc(`${uid}/journal/notes/${id}`).delete()
  
      dispatch(deleteNote(id))

     }


}

export const deleteNote =(id) =>({
    type: types.notesDelete,
    payload: id

})

export const noteLogout = () => ({ //borrar las notas de memoria cuando el usuario se desloguee
     type: types.notesLogoutCleaning,
     payload: []
})