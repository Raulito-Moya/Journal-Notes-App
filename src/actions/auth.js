import Swal from 'sweetalert2'
import withReactContent from 'sweetalert2-react-content'

import {firebase, googleAuthProvider} from '../firebase/firebase-config';
import { types } from "../types/types";
import { noteLogout } from './notes';
import { finishLoading, startLoading } from './ui';
 
const MySwal = withReactContent(Swal) //para los efectos de aniamacion en el form

export const startLoginEmailPassword = (email,password) => { //aqui voy a simular una accion asincrona
     return (dispatch) => {
        
       dispatch(startLoading())
     return  firebase.auth().signInWithEmailAndPassword(email,password) //esta instruccion es para logear al usuario
              .then( ({user}) => {
                 // console.log(user)
               dispatch( login(user.uid, user.displayName))  
               dispatch(finishLoading())
            })
            .catch(e => {
               console.log(e); 
               dispatch(finishLoading()) //por si hay un error
               MySwal.fire( 'Error', e.message , 'error')
              })

     }

};

export const startRegisterRegisterWithEmailPasswordName = ( email, password, name) => {

   return( dispatch ) => {

      firebase.auth().createUserWithEmailAndPassword(email, password)
            .then( async({user}) =>{

            await user.updateProfile({ displayName: name }) //update user profile data, se espera que se actualize el objeto para despues hacr el dispatch

            //  console.log(user);
                  dispatch( 
                       login(user.uid, user.displayName) 
                      ) 
             })
             .catch(e =>{
               console.log(e);
               MySwal.fire( 'Error', e.message , 'error')
             })
   }



};

export const startGoogleLogin = () => {  //aqui hago un sign in con google
   return (dispatch) => {
  
     firebase.auth().signInWithPopup(googleAuthProvider)
           .then( ({user}) =>{
             dispatch( login(user.uid, user.displayName) ) 
           })
   }


};

export const login = (uid, displayName) => ({
       type: types.login,
       payload: {
         uid,
         displayName
       }
});


export const startLogout = () => {
     return async( dispatch ) => {

      await firebase.auth().signOut();
      dispatch(logout())
      dispatch(noteLogout()) //
     }

}


export const logout = () => ({
   type: types.logout

})