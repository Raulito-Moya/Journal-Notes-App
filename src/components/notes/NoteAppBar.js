import React from 'react'
import { useDispatch, useSelector } from 'react-redux';
import { startSaveNote, startUpLoading } from '../../actions/notes'

export const NoteAppBar = () => {
  
  const dispatch = useDispatch();
  const {active} = useSelector(state => state.notes)


  const handleSave = () => {
              console.log(active);
    dispatch(startSaveNote(active))
  }

  const handlePictureClick = () =>{
      console.log('picture');
      document.querySelector('#fileSelector').click()
 
  }


  const handleFileChange = (e) => {
     // console.log(e.target.files);
      const file = e.target.files[0]
    if( file ){
       dispatch(startUpLoading(file))
    }
  }

  return(
       <div className="notes__appbar">
            <span>28 de agosto 2000</span>

          <input
             id="fileSelector" 
             type="file" 
             style={{display:'none'}}
             name="file" 
             onChange={handleFileChange}/>
     
            <div>
               <button 
                  type="click" 
                   className="btn"
                   onClick={handlePictureClick}>
                   picture
                </button>

                <button 
                    type="click" 
                     className="btn"
                      onClick={handleSave}
                      >
                    save
                </button>
            </div>
       </div>
  )
 
}