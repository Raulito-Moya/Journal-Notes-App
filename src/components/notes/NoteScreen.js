import React, { useEffect, useRef } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { activeNote, startDeleting } from '../../actions/notes'
import { useForm } from '../../hooks/useForm'
import { NoteAppBar } from './NoteAppBar'



export const NoteScreen = () => {

  const {active:note} = useSelector( state => state.notes)

   const [ formValues, handleInputchange ,reset] =  useForm( note );
   const {body, title, id} = formValues;
 
   const activeId = useRef( note.id ) //guarda la referencia de note.id
  
   const dispatch = useDispatch();

      useEffect(()=>{

         if(note.id !== activeId.current){
           reset(note);
          activeId.current = note.id
         }
      
      },[note,reset])
  

      useEffect(()=>{ //actualizar la nota activa

        dispatch(activeNote(formValues.id, {...formValues}))


      },[formValues, dispatch])




   const handleDelete = () => {
     dispatch(startDeleting( id ))
   }


  return(
      <div className="notes__main-content">
          
         <NoteAppBar/>
   
       <div className="notes__content">
           
           <input 
             type="text"
             className="notes__title-input"
              autoComplete="off" 
              placeholder='Some awesome title'
              name="title"
               value={title}
               onChange={handleInputchange}
              />

              <textarea 
                placeholder="Whats happends today?"
                 className="notes__textarea" 
                 name="body"
                 value={body}
                 onChange={handleInputchange}>
               
            </textarea>

           {
             (note.url) 
               &&  (
                <div className="notes__image">
                   <img src={ note.url }
                      alt="imagen"
                   />
                </div>
                )
           } 

            
       </div>

     <button 
       type="click"
       className="btn btn-danger"
       onClick={handleDelete}
       >
        Delete
       </button>

      </div>
  )
 


}