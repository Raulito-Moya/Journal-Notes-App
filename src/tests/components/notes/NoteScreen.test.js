import { mount } from 'enzyme';
import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import {MemoryRouter} from 'react-router-dom';
import { activeNote } from '../../../actions/notes';
import { NoteScreen } from '../../../components/notes/NoteScreen';

jest.mock('../../../actions/notes',()=>({
   activeNote: jest.fn()
}))


const middlewares = [thunk]
const mockStore = configureStore(middlewares)


const initialstate = {
    auth: {
        uid: '1',
        name: 'Fernando'
    },
    ui: {
       loading: false,
       msgError: null
    },
    notes: {
        active: {
            id: 1234,
            title: 'Hola',
            body: 'Mundo',
            date: 0
            
        },
        notes: []
    }
 }
 
    let store = mockStore(initialstate) 
     store.dispatch = jest.fn() //vamos aremplazar la funcionalidad que tenga el store por una funcion
 
   
 const wrapper = mount(
         <Provider store={store}>
           <NoteScreen/>
         </Provider>
     )

describe('Preubas en el <NoteScreen/>', () => {


  test('debe de mostararse correctamente',() => { 
  
    expect( wrapper ).toMatchSnapshot()
  
  })


  test('debe de disparar el activeNotes',()=>{

    wrapper.find('input[name="title"]').simulate('change', {
        target: {
            name: 'title',
            value: 'Hola de nuevo'
        }
    });
         
    expect( activeNote ).toHaveBeenLastCalledWith(
        1234,
        {
            body: 'Mundo',
            title: 'Hola de nuevo',
            id: 1234,
            date: 0
        }
    )

  })

});