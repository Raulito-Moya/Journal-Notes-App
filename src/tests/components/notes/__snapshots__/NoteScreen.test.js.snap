// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Preubas en el <NoteScreen/> debe de mostararse correctamente 1`] = `
<div
  className="notes__main-content"
>
  <div
    className="notes__appbar"
  >
    <span>
      28 de agosto 2000
    </span>
    <input
      id="fileSelector"
      name="file"
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
    />
    <div>
      <button
        className="btn"
        onClick={[Function]}
        type="click"
      >
        picture
      </button>
      <button
        className="btn"
        onClick={[Function]}
        type="click"
      >
        save
      </button>
    </div>
  </div>
  <div
    className="notes__content"
  >
    <input
      autoComplete="off"
      className="notes__title-input"
      name="title"
      onChange={[Function]}
      placeholder="Some awesome title"
      type="text"
      value="Hola"
    />
    <textarea
      className="notes__textarea"
      name="body"
      onChange={[Function]}
      placeholder="Whats happends today?"
      value="Mundo"
    />
  </div>
  <button
    className="btn btn-danger"
    onClick={[Function]}
    type="click"
  >
    Delete
  </button>
</div>
`;
